### ISSUE TITLE
Benchmarks on M1 Max

### ISSUE THREAD

**Issue by awni** [mlx-graphs/mlx-graphs#92]
Benchmarking MLX vs PT graphs on an M1 Max:

Findings:
1 . MLX is faster on every case compared to PT
2. Compiling helps a lot but only on the bigger dataset (it's 2x in one case on the bigger dataset 🚀 )
3. Suspect issues with smaller graphs are in the scatter still just based on some light profiling

### PT:
```
BZR_MD
==========
GCNConv | 0.190s
GATConv | 0.312s

MUTAG
==========
Downloading https://www.chrsmrrs.com/graphkerneldatasets/MUTAG.zip
Extracting .mlx_graphs_data/MUTAG/MUTAG/MUTAG.zip
Processing...
Done!
GCNConv | 0.046s
GATConv | 0.067s

DD
==========
Downloading https://www.chrsmrrs.com/graphkerneldatasets/DD.zip
Extracting .mlx_graphs_data/DD/DD/DD.zip
Processing...
Done!
GCNConv | 3.930s
GATConv | 5.088s
```

### MLX No Compile:

```
BZR_MD
==========
GCNConv | 0.088s
GATConv | 0.114s

MUTAG
==========
GCNConv | 0.027s
GATConv | 0.035s

DD
==========
GCNConv | 2.026s
GATConv | 3.027s
```

### MLX Compiled

```
BZR_MD
==========
GCNConv | 0.083s
GATConv | 0.111s

MUTAG
==========
GCNConv | 0.025s
GATConv | 0.033s

DD
==========
GCNConv | 1.267s
GATConv | 1.513s
```

**Comment by TristanBilot** [comment]
Thanks for the benchmark @awni, this is very interesting!
The M1 Max is clearly a game changer here compared to the M1 Pro and M3 Pro we already benchmarked. How many GPU cores and RAM do you have?

The compiled functions on the large graph are very neat, it seems clearly worth it to keep compiled functions by default in the framework, what do you think @awni ?

I think those results are sufficient for a first official release 🚀

**Comment by awni** [comment]
>  it seems clearly worth it to keep compiled functions by default in the framework, what do you think @awni ?

For MLX Graphs I do think you should probably default compile the graph training step update. The only reason not to is if the input sizes change shape from batch to batch but it does not look like that is the case for you.

For MLX we want to default compile but compiling can be slow when the input shapes change so we are working on a shapeless compile which will work at least for a large number of functions that we want to default compile: https://github.com/ml-explore/mlx/pull/687

> How many GPU cores and RAM do you have?

```
  Model Name:	MacBook Pro
  Chip:	Apple M1 Max
  Total Number of CPU Cores:	10 (8 performance and 2 efficiency)
  Memory:	32 GB
  Total Number of GPU Cores:	32
```

**Comment by awni** [comment]
```
dgl | GCNConv | 0.130s
dgl | GATConv | 0.340s

pyg | GCNConv | 0.241s
pyg | GATConv | 0.341s

Downloading /Users/awnihannun/repos/thirdparty/mlx-graphs/docs/source/tutorials/notebooks/.mlx_graphs_data/BZR_MD/raw/BZR_MD.zip from https://www.chrsmrrs.com/graphkerneldatasets/BZR_MD.zip...
mxg | GCNConv | 0.024s
mxg | GATConv | 0.113s

MUTAG
==========
Downloading https://www.chrsmrrs.com/graphkerneldatasets/MUTAG.zip
Extracting .mlx_graphs_data/MUTAG/MUTAG/MUTAG.zip
Processing...
Done!
Downloading /Users/awnihannun/.dgl/MUTAG.zip from https://www.chrsmrrs.com/graphkerneldatasets/MUTAG.zip...
/Users/awnihannun/.dgl/MUTAG.zip: 100%|███████████████████████████████████████████████████████████████████████████████████| 24.6k/24.6k [00:00<00:00, 170kB/s]
Extracting file to /Users/awnihannun/.dgl/MUTAG_47395044
dgl | GCNConv | 0.036s
dgl | GATConv | 0.063s

pyg | GCNConv | 0.052s
pyg | GATConv | 0.080s

Downloading /Users/awnihannun/repos/thirdparty/mlx-graphs/docs/source/tutorials/notebooks/.mlx_graphs_data/MUTAG/raw/MUTAG.zip from https://www.chrsmrrs.com/graphkerneldatasets/MUTAG.zip...
mxg | GCNConv | 0.016s
mxg | GATConv | 0.036s

DD
==========
Downloading https://www.chrsmrrs.com/graphkerneldatasets/DD.zip
Extracting .mlx_graphs_data/DD/DD/DD.zip
Processing...
Done!
Downloading /Users/awnihannun/.dgl/DD.zip from https://www.chrsmrrs.com/graphkerneldatasets/DD.zip...
/Users/awnihannun/.dgl/DD.zip: 100%|█████████████████████████████████████████████████████████████████████████████████████| 4.98M/4.98M [00:01<00:00, 3.56MB/s]
Extracting file to /Users/awnihannun/.dgl/DD_a6864905
dgl | GCNConv | 2.243s
dgl | GATConv | 4.838s

pyg | GCNConv | 4.176s
pyg | GATConv | 5.380s

Downloading /Users/awnihannun/repos/thirdparty/mlx-graphs/docs/source/tutorials/notebooks/.mlx_graphs_data/DD/raw/DD.zip from https://www.chrsmrrs.com/graphkerneldatasets/DD.zip...
mxg | GCNConv | 0.743s
mxg | GATConv | 1.535s
```

**Comment by awni** [comment]
Maybe it's time to get a Max ;)

**Comment by francescofarina** [comment]
Amazing, thanks @awni - we'll include those numbers in the results. 

We still see some amount of variability in the benchmarks (for all frameworks and platforms), but the trend is pretty clear with mlx emerging as the winner.

**Comment by francescofarina** [comment]
@awni it would be interesting to get your results also on a larger dataset such as NCI-H23 to see whether any speedup we get by compiling is consistent. On my machine I actually see that compiling the training step results in a speedup only with smaller datasets, on larger one I get a significant performance drop - see [here](https://github.com/mlx-graphs/mlx-graphs/blob/d1b58593c21b277844d9c4b7f14d1d50c5c28fad/benchmarks/framework_comparison/training_loop/results/M3Pro_5P_6E_14GPU_18G.md). You can run the benchmark with [this script](https://github.com/mlx-graphs/mlx-graphs/blob/d1b58593c21b277844d9c4b7f14d1d50c5c28fad/benchmarks/framework_comparison/training_loop/launcher.py).

I did some profiling [here](https://github.com/mlx-graphs/mlx-graphs/blob/4a47e4fb3b00680e13f9164e71024ca913e0e039/benchmarks/profiling/profile_training.py) and it seems that the extra time I see in the compiled code all goes to `mlx.core.eval`:

- This is the profiling with a non-compiled training step
<img width="1590" alt="Screenshot 2024-02-19 at 10 00 22" src="https://github.com/mlx-graphs/mlx-graphs/assets/19568231/7c987eef-3433-407f-a1dd-6a8a13fddbaa">

- This is the profiling with the compiled one
<img width="1591" alt="Screenshot 2024-02-19 at 09 59 11" src="https://github.com/mlx-graphs/mlx-graphs/assets/19568231/8f8fabc8-189b-4ecf-a5fc-f83ed122451b">

I'm still not sure what's causing it, so any extra runs are welcome.

**Comment by awni** [comment]
I will check. Also just FYI we have a `mx.disable_compile()` and `mx.enable_compile()` as it's a common use case to want to turn it off and on for debugging / perf etc. So you don't need to pass compile flags around and duplicate cade. (There is also the `MLX_DISABLE_COMPILE` environment variable.

**Comment by awni** [comment]
It's also slower on my M1 ... not sure why but that's something to investigate.

```
| Dataset | Framework | Layer | Time/epoch |
| --- | --- | --- | --- |
| NCI-H23 | mxg | GCNConv | 4.097s |
| NCI-H23 | mxg | GATConv | 5.620s |
| NCI-H23 | mxg(compiled) | GCNConv | 6.434s |
| NCI-H23 | mxg(compiled) | GATConv | 7.131s |
```

**Comment by awni** [comment]
Here's the other benchmarks using that script:

| Dataset | Framework | Layer | Time/epoch |
| --- | --- | --- | --- |
| BZR_MD | dgl | GCNConv | 0.062s |
| BZR_MD | dgl | GATConv | 0.164s |
| BZR_MD | pyg | GCNConv | 0.109s |
| BZR_MD | pyg | GATConv | 0.162s |
| BZR_MD | mxg | GCNConv | 0.051s |
| BZR_MD | mxg | GATConv | 0.062s |
| BZR_MD | mxg(compiled) | GCNConv | 0.039s |
| BZR_MD | mxg(compiled) | GATConv | 0.051s |
| MUTAG | dgl | GCNConv | 0.016s |
| MUTAG | dgl | GATConv | 0.029s |
| MUTAG | pyg | GCNConv | 0.023s |
| MUTAG | pyg | GATConv | 0.032s |
| MUTAG | mxg | GCNConv | 0.017s |
| MUTAG | mxg | GATConv | 0.021s |
| MUTAG | mxg(compiled) | GCNConv | 0.010s |
| MUTAG | mxg(compiled) | GATConv | 0.013s |
| DD | dgl | GCNConv | 1.090s |
| DD | dgl | GATConv | 2.375s |
| DD | pyg | GCNConv | 2.038s |
| DD | pyg | GATConv | 2.637s |
| DD | mxg | GCNConv | 1.276s |
| DD | mxg | GATConv | 1.366s |
| DD | mxg(compiled) | GCNConv | 0.605s |
| DD | mxg(compiled) | GATConv | 0.722s |

**Comment by francescofarina** [comment]
Ok, that's interesting @awni. Looks like on M1 Max you have a performance drop only on NCI-H23, while on M3 Pro I have it both on NCI-H23 and DD. I have both less GPU cores (14 VS 32) and RAM (18 VS 32) - memory doesn't seem to be an issue, so I'm wondering whether there's something going on between how we perform message passing and GPU usage.

**Comment by awni** [comment]
Benchmarks on M2 Ultra:

| Dataset | Framework | Layer | Time/epoch |
| --- | --- | --- | --- |
| BZR_MD | dgl | GCNConv | 0.053s |
| BZR_MD | dgl | GATConv | 0.148s |
| BZR_MD | pyg | GCNConv | 0.081s |
| BZR_MD | pyg | GATConv | 0.120s |
| BZR_MD | mxg | GCNConv | 0.034s |
| BZR_MD | mxg | GATConv | 0.041s |
| BZR_MD | mxg(compiled) | GCNConv | 0.020s |
| BZR_MD | mxg(compiled) | GATConv | 0.026s |
| MUTAG | dgl | GCNConv | 0.014s |
| MUTAG | dgl | GATConv | 0.028s |
| MUTAG | pyg | GCNConv | 0.025s |
| MUTAG | pyg | GATConv | 0.039s |
| MUTAG | mxg | GCNConv | 0.013s |
| MUTAG | mxg | GATConv | 0.017s |
| MUTAG | mxg(compiled) | GCNConv | 0.008s |
| MUTAG | mxg(compiled) | GATConv | 0.010s |
| DD | dgl | GCNConv | 0.815s |
| DD | dgl | GATConv | 1.912s |
| DD | pyg | GCNConv | 2.779s |
| DD | pyg | GATConv | 3.403s |
| DD | mxg | GCNConv | 0.475s |
| DD | mxg | GATConv | 0.522s |
| DD | mxg(compiled) | GCNConv | 0.239s |
| DD | mxg(compiled) | GATConv | 0.284s |
| NCI-H23 | dgl | GCNConv | 5.306s |
| NCI-H23 | dgl | GATConv | 9.349s |
| NCI-H23 | pyg | GCNConv | 10.864s |
| NCI-H23 | pyg | GATConv | 15.793s |
| NCI-H23 | mxg | GCNConv | 3.168s |
| NCI-H23 | mxg | GATConv | 3.978s |
| NCI-H23 | mxg(compiled) | GCNConv | 3.880s |
| NCI-H23 | mxg(compiled) | GATConv | 4.304s |

**Comment by awni** [comment]
@francescofarina I notice that the arrays in the `graph` data batch are changing size from iteration to iteration. This could explain the slowdown we observe with compile. Right now if you change the shapes then the function will be recompiled which can be expensive for large functions (see [docs](https://ml-explore.github.io/mlx/build/html/usage/compile.html)).

**Comment by francescofarina** [comment]
> I was going to say this feels like it could be a memory use thing..but I'm not certain.

So, on M3 Pro I see that with DD I have both full RAM usage and GPU usage. With NCI-H23 only full GPU usage.

> Even on the ultra NCI is slower compiled which is interesting.. can you share more about what is different from that dataset and DD? Is it just the size?

- DD 
```
Dataset attributes
--------------------
Number of graphs: 1178
Number of node features: 89
Number of edge features: 0
Number of graph features: 0
Number of graph classes to predict: 2

Dataset stats
--------------------
Mean node degree: 4.98
Mean num of nodes: 284.32
Mean num of edges: 1431.32
```
- NCI-H23 
```
Dataset attributes
--------------------
Number of graphs: 40353
Number of node features: 65
Number of edge features: 3
Number of graph features: 0
Number of graph classes to predict: 2

Dataset stats
--------------------
Mean node degree: 2.15
Mean num of nodes: 26.07
Mean num of edges: 56.20
```

**Comment by francescofarina** [comment]
> @francescofarina I notice that the arrays in the `graph` data batch are changing size from iteration to iteration. This could explain the slowdown we observe with compile. Right now if you change the shapes then the function will be recompiled which can be expensive for large functions (see [docs](https://ml-explore.github.io/mlx/build/html/usage/compile.html)).

Yes, you're right, each graph is different so when batching the resulting size may change.

**Comment by awni** [comment]
> Yes, you're right, each graph is different so when batching the resulting size may change.

I see.. in that case we may not want to use compile except for the smaller graphs where there is only one or a small number of batches.

An alternative is to try to either pad to the same size or to a few fixed sizes. But I don't know how that would go in your data loader and if it makes sense.

**Comment by francescofarina** [comment]
Yes, that makes sense - I guess we investigate padding each graph in a batch to match the dimension of the largest one in the dataset so that the size stays constant, @TristanBilot wdyt?

It's interesting though how compiling could already make sense anyways for some datasets depending on the specific platform.

Thanks @awni for the support on this point!

**Comment by awni** [comment]
@francescofarina I'm not certain that's the whole story. It would be good to verify that assumption with e.g. a fixed size batch to take that out of the equation.

**Comment by TristanBilot** [comment]
Thanks @awni @francescofarina for your benchmarks. Very interesting to test different scenarios. 

I find surprising that the bottleneck comes from the different batch shapes, because the shapes of each batch in the other datasets are also always different. Contrary to "standard" batching, we don't add a fixed-size batch dimension to the batch but we stack the edges from multiple graphs together on the 0 axis. As the graphs have an arbitrary number of edges, it will be always different, except if we use padding as mentioned by @francescofarina. 

@awni, to avoid re-compiling, all the inputs given to the compiled function should have the same shape right? In this case we should pad both the indices and the features @francescofarina. If we do so it may be faster for the other datasets too, no?

For the NCI-H23 dataset, I think the bottleneck may come from the very high number of graphs compared to the other datasets. This results in many batches and re-compilations, so maybe we can measure runtime for a same number of batches as in another smaller dataset?

**Comment by awni** [comment]
Actually looking at your benchmark I'm not certain:

```
                times = timeit.Timer(
                    lambda: benchmark(framework, loader, step, state)
                ).repeat(repeat=TIMEIT_REPEAT, number=TIMEIT_NUMBER)

                time = min(times) / TIMEIT_NUMBER
```

Assuming that the batch shapes don't change from training run to training run, the first iteration of the timer might be really slow because it is compiling for all the different shapes. But then after that it should be fast. 

Maybe that assumption is wrong?

If it's not then this also suggests there is a deeper issue with the bigger NCI dataset...

**Comment by francescofarina** [comment]
No, the data loader has `shuffle=True` so for each epoch (or each new repeat of timeit) the batches are potentially different. That's a good way to check the assumption though, by setting `shuffle=False`. I'll do that.

**Comment by francescofarina** [comment]
So, interesting enough, by setting `shuffle=False`, the compiled version still takes much longer. Even by setting `EPOCHS` > 1. So there's probably a deeper issue than not having fixed-size inputs. 
Now, the really weird part is that If I use only the first ~1000 samples of the NCI-H23 dataset, so that it has roughly the same size of the DD dataset, the performance of the compiled version is stil ~4x slower, while for DD it's almost the same. And DD contains pretty larger graphs as well (see https://github.com/mlx-graphs/mlx-graphs/issues/92#issuecomment-1952751386). @TristanBilot could you please run the same experiment?

**Comment by TristanBilot** [comment]
Here are my results on M1 Pro (without the new scatter):
```
Platform macOS-14.2
mlx version: 0.3.0
mlx-graphs version: 0.0.2
torch version: 2.1.2
torch_geometric version: 2.4.0
dgl version: 2.0.0

| Dataset | Framework | Layer | Time/epoch |
| --- | --- | --- | --- |
| NCI-H23 | dgl | GCNConv | 5.046s |
| NCI-H23 | dgl | GATConv | 8.803s |
| NCI-H23 | pyg | GCNConv | 7.253s |
| NCI-H23 | pyg | GATConv | 10.505s |
| NCI-H23 | mxg | GCNConv | 5.585s |
| NCI-H23 | mxg | GATConv | 7.601s |
| NCI-H23 | mxg(compiled) | GCNConv | 11.256s |
| NCI-H23 | mxg(compiled) | GATConv | 12.603s |
| BZR_MD | dgl | GCNConv | 0.066s |
| BZR_MD | dgl | GATConv | 0.174s |
| BZR_MD | pyg | GCNConv | 0.130s |
| BZR_MD | pyg | GATConv | 0.175s |
| BZR_MD | mxg | GCNConv | 0.101s |
| BZR_MD | mxg | GATConv | 0.113s |
| BZR_MD | mxg(compiled) | GCNConv | 0.062s |
| BZR_MD | mxg(compiled) | GATConv | 0.083s |
| MUTAG | dgl | GCNConv | 0.017s |
| MUTAG | dgl | GATConv | 0.032s |
| MUTAG | pyg | GCNConv | 0.025s |
| MUTAG | pyg | GATConv | 0.037s |
| MUTAG | mxg | GCNConv | 0.024s |
| MUTAG | mxg | GATConv | 0.031s |
| MUTAG | mxg(compiled) | GCNConv | 0.013s |
| MUTAG | mxg(compiled) | GATConv | 0.017s |
| DD | dgl | GCNConv | 1.079s |
| DD | dgl | GATConv | 2.486s |
| DD | pyg | GCNConv | 2.186s |
| DD | pyg | GATConv | 2.889s |
| DD | mxg | GCNConv | 1.413s |
| DD | mxg | GATConv | 1.481s |
| DD | mxg(compiled) | GCNConv | 1.002s |
| DD | mxg(compiled) | GATConv | 1.235s |
```

I will try to pad the batches to the same shape to see if it improves.

**Comment by awni** [comment]
> And DD contains pretty larger graphs

That's interesting.. I wonder why that would make a difference 🤔

**Comment by awni** [comment]
It does seem like there is an issue with compile with those sizes.. we'll need to investigate further.

**Comment by TristanBilot** [comment]
With padding, I end up with very similar results for NCI-H23. Is it possible that compiled functions from a previous run are also used in other runs @awni ? Are compiled functions re-compiled at the beginning of each script or are they kind of stored on disk and living across multiple script runs?

| Dataset | Framework | Layer | Time/epoch |
| --- | --- | --- | --- |
| BZR_MD | mxg | GCNConv | 0.147s |
| BZR_MD | mxg | GATConv | 0.242s |
| BZR_MD | mxg(compiled) | GCNConv | 0.056s |
| BZR_MD | mxg(compiled) | GATConv | 0.078s |
| MUTAG | mxg | GCNConv | 0.025s |
| MUTAG | mxg | GATConv | 0.028s |
| MUTAG | mxg(compiled) | GCNConv | 0.013s |
| MUTAG | mxg(compiled) | GATConv | 0.016s |
| DD | mxg | GCNConv | 1.677s |
| DD | mxg | GATConv | 1.860s |
| DD | mxg(compiled) | GCNConv | 1.426s |
| DD | mxg(compiled) | GATConv | 1.760s |
| NCI-H23 | mxg | GCNConv | 5.889s |
| NCI-H23 | mxg | GATConv | 7.827s |
| NCI-H23 | mxg(compiled) | GCNConv | 3.205s |
| NCI-H23 | mxg(compiled) | GATConv | 4.358s |

**Comment by francescofarina** [comment]
@TristanBilot any changes in your local environment that could have affected it?

**Comment by awni** [comment]
>  Is it possible that compiled functions from a previous run are also used in next runs @awni ? Are compiled functions re-compiled at the beginning of each script or are they kind of stored on disk and living across multiple script runs?

No the compiled function is not stored from run to run.

I wonder if there is a bug with how mx.compile is interacting with your padding. Did you compile through the padding code?

**Comment by TristanBilot** [comment]
I opened a draft PR with the padded loader I used for experiments.

> @TristanBilot any changes in your local environment that could have affected it?

No IMO, I'm pretty sure I didn't change anything in the env.

> I wonder if there is a bug with how mx.compile is interacting with your padding. Did you compile through the padding code?

I called `compile` directly from the `step` function used in the training loop, as in previous benchmarks.

**Comment by TristanBilot** [comment]
it would be interesting if you re-run the exps with the original loader and padded loader @francescofarina (just comment/uncomment `loader = mxg_loaders.Dataloader(dataset, batch_size=batch_size, shuffle=True)`)

**Comment by TristanBilot** [comment]
btw, as we batch the graph w.r.t the number of edges in the graph instead of the number of graphs, I compute an approximate batch size with `mean_nb_edges_in_dataset * BATCH_SIZE`.

**Comment by francescofarina** [comment]
So, on your branch I also get better results with the compiled version, no matter what.

With the new `PaddedDataloader` I get

| Dataset | Framework | Layer | Time/epoch |
| --- | --- | --- | --- |
| BZR_MD | mxg | GCNConv | 0.043s |
| BZR_MD | mxg | GATConv | 0.059s |
| BZR_MD | mxg(compiled) | GCNConv | 0.036s |
| BZR_MD | mxg(compiled) | GATConv | 0.050s |
| MUTAG | mxg | GCNConv | 0.011s |
| MUTAG | mxg | GATConv | 0.014s |
| MUTAG | mxg(compiled) | GCNConv | 0.007s |
| MUTAG | mxg(compiled) | GATConv | 0.009s |
| DD | mxg | GCNConv | 0.504s |
| DD | mxg | GATConv | 0.769s |
| DD | mxg(compiled) | GCNConv | 0.456s |
| DD | mxg(compiled) | GATConv | 0.648s |
| NCI-H23 | mxg | GCNConv | 3.136s |
| NCI-H23 | mxg | GATConv | 4.107s |
| NCI-H23 | mxg(compiled) | GCNConv | 1.939s |
| NCI-H23 | mxg(compiled) | GATConv | 2.515s |

With the standard one

| Dataset | Framework | Layer | Time/epoch |
| --- | --- | --- | --- |
| BZR_MD | mxg | GCNConv | 0.046s |
| BZR_MD | mxg | GATConv | 0.062s |
| BZR_MD | mxg(compiled) | GCNConv | 0.041s |
| BZR_MD | mxg(compiled) | GATConv | 0.057s |
| MUTAG | mxg | GCNConv | 0.011s |
| MUTAG | mxg | GATConv | 0.014s |
| MUTAG | mxg(compiled) | GCNConv | 0.007s |
| MUTAG | mxg(compiled) | GATConv | 0.010s |
| DD | mxg | GCNConv | 0.588s |
| DD | mxg | GATConv | 0.814s |
| DD | mxg(compiled) | GCNConv | 0.553s |
| DD | mxg(compiled) | GATConv | 0.793s |
| NCI-H23 | mxg | GCNConv | 2.929s |
| NCI-H23 | mxg | GATConv | 3.884s |
| NCI-H23 | mxg(compiled) | GCNConv | 2.045s |
| NCI-H23 | mxg(compiled) | GATConv | 2.707s |

This is weird as the only change you made seems to be loading the dataset just once per framework
<img width="647" alt="Screenshot 2024-02-21 at 16 49 31" src="https://github.com/mlx-graphs/mlx-graphs/assets/19568231/2675a232-4869-45d0-a20d-4e77f7f4374b">

**Comment by TristanBilot** [comment]
> This is weird as the only change you made seems to be loading the dataset just once per framework

Haha yes I thought of this one, do you think this can change anything? It's worth trying just in case 😅

**Comment by francescofarina** [comment]
If I revert that I go back to the previous results - very weird.

**Comment by TristanBilot** [comment]
Ok I found out. Indeed, it's when you change the dataset line that we get back to the original results.

**Comment by TristanBilot** [comment]
I don't get why this changes the results given that we use `shuffle=True` though. Also, the padding seems yield similar results as the original loader on NCI-H23.

**Comment by TristanBilot** [comment]
Ok, it seems that we only shuffle the batch when we reach the end of the iterator. We don't shuffle the first epoch of the batch. Our experiments run only for one epoch so the batches are basically always the same...

**Comment by francescofarina** [comment]
but we shuffle the indices in the dataloader, not the dataset itself, so I'm not sure why that's a problem.

**Comment by TristanBilot** [comment]
What's weird is that it's faster with compile when we use the same dataset for both the non-compiled and compiled steps. This means the first run with `compile=False` uses no shuffle, and the second run with `compile=True` uses shuffle. So we get better performance when the shapes are different 🤔

**Comment by francescofarina** [comment]
So, new update - even by loading the dataset in the outer loop as in @TristanBilot's PR, if the non-compiled version runs before the compiled one we get better results, however if the compiled version runs before the non-compiled one then we're back to the old case where the compiled version is about 4x slower.

- compiled first:
```
| NCI-H23 | mxg(compiled) | GCNConv | 12.909s |
| NCI-H23 | mxg(compiled) | GATConv | 13.457s |
| NCI-H23 | mxg | GCNConv | 3.221s |
| NCI-H23 | mxg | GATConv | 4.434s |
```

- non-compiled first

```
| NCI-H23 | mxg | GCNConv | 3.269s |
| NCI-H23 | mxg | GATConv | 4.667s |
| NCI-H23 | mxg(compiled) | GCNConv | 2.235s |
| NCI-H23 | mxg(compiled) | GATConv | 2.796s |
```

**Comment by awni** [comment]
Can you give me a way to repro that?

**Comment by francescofarina** [comment]
On this branch https://github.com/mlx-graphs/mlx-graphs/tree/batch_padding, change the order of `True, False` [here](https://github.com/mlx-graphs/mlx-graphs/blob/8735dc5ea8e3a2f70b7a186715ea47684ffc1883/benchmarks/framework_comparison/training_loop/launcher.py#L37) to specify whether to run the compiled one first or second. And you can select the dataloader to use [here](https://github.com/mlx-graphs/mlx-graphs/blob/8735dc5ea8e3a2f70b7a186715ea47684ffc1883/benchmarks/framework_comparison/training_loop/mlx_graphs_setup.py#L48)

**Comment by awni** [comment]
I uncovered a couple of problems with using `mx.compile` in your current code:

1. You currently use a data class for the [graph data](https://github.com/mlx-graphs/mlx-graphs/blob/main/mlx_graphs/data/data.py#L7). This will not be tracked by `mx.compile` so the inputs are effectively constant. Basically all of the benchmarks using `mx.compile` thus far are not valid :\. 
2. If you switch the graph data to a dictionary then it will crash because of the [eval inside the step function](https://github.com/mlx-graphs/mlx-graphs/blob/main/mlx_graphs/nn/pooling/global_pooling.py#L28)

We can fix this by sending the `step` function the dictionary  `GraphData.dict` rather than the object itself.  And then replacing something like:

```
    out_size = batch_indices.max().item() + 1
```

with something like `node_features.shape[-1]`

**Comment by TristanBilot** [comment]
How should we know that compile didn't work here? Why is there no error message? I replaced the `GraphData` by a dict with `to_dict()` and also computed `out_size = batch_indices.max().item() + 1` in the data loader so that we don't have to eval in `step()`.

However, I get those results using the original data loader:

| Dataset | Framework | Layer | Time/epoch |
| --- | --- | --- | --- |
| BZR_MD | mxg | GCNConv | 0.069s |
| BZR_MD | mxg | GATConv | 0.092s |
| BZR_MD | mxg(compiled) | GCNConv | 0.070s |
| BZR_MD | mxg(compiled) | GATConv | 0.094s |
| MUTAG | mxg | GCNConv | 0.023s |
| MUTAG | mxg | GATConv | 0.027s |
| MUTAG | mxg(compiled) | GCNConv | 0.023s |
| MUTAG | mxg(compiled) | GATConv | 0.030s |
| DD | mxg | GCNConv | 1.320s |
| DD | mxg | GATConv | 1.482s |
| DD | mxg(compiled) | GCNConv | 1.133s |
| DD | mxg(compiled) | GATConv | 1.357s |
| NCI-H23 | mxg | GCNConv | 5.321s |
| NCI-H23 | mxg | GATConv | 7.044s |
| NCI-H23 | mxg(compiled) | GCNConv | 4.710s |
| NCI-H23 | mxg(compiled) | GATConv | 6.683s |

And those using the padded loader:

| Dataset | Framework | Layer | Time/epoch |
| --- | --- | --- | --- |
| BZR_MD | mxg | GCNConv | 0.156s |
| BZR_MD | mxg | GATConv | 0.254s |
| BZR_MD | mxg(compiled) | GCNConv | 0.138s |
| BZR_MD | mxg(compiled) | GATConv | 0.187s |
| MUTAG | mxg | GCNConv | 0.022s |
| MUTAG | mxg | GATConv | 0.033s |
| MUTAG | mxg(compiled) | GCNConv | 0.023s |
| MUTAG | mxg(compiled) | GATConv | 0.029s |
| DD | mxg | GCNConv | 1.792s |
| DD | mxg | GATConv | 2.163s |
| DD | mxg(compiled) | GCNConv | 1.451s |
| DD | mxg(compiled) | GATConv | 1.651s |
| NCI-H23 | mxg | GCNConv | 6.064s |
| NCI-H23 | mxg | GATConv | 8.010s |
| NCI-H23 | mxg(compiled) | GCNConv | 5.561s |
| NCI-H23 | mxg(compiled) | GATConv | 7.484s |

Am I missing something? I pushed the code on the branch https://github.com/mlx-graphs/mlx-graphs/pull/101.

**Comment by awni** [comment]
> How should we know that compile didn't work here? 

That's a great question. It's not at all on you. It's on us to figure out how to catch that and give the proper feedback. Maybe we should disallow non-standard objects 🤔 , I'm not sure yet. Right now it fails silently though which is no good 😞 . 

You would definitely notice outside of a benchmark because you should be getting bogus results. As a start, I would verify that all of these additions are actually producing the right result.

**Comment by awni** [comment]
Regarding your new results:

> Am I missing something?

Can you elaborate? What would you be missing?

**Comment by TristanBilot** [comment]
> It's on us to figure out how to catch that and give the proper feedback. Maybe we should disallow non-standard objects 🤔

Yeah it would be useful to either disallow or warn the user in such cases.

> You would definitely notice outside of a benchmark because you should be getting bogus results.

Yes it makes sense.

> Can you elaborate? What would you be missing?

In the last results, there's almost no difference between non-compiled and compiled versions (it's even worse on small datasets). Padding does not seem to affect the performance as well. Do you think these are the actual expected results or `compile` may fail to be properly applied here too?
@francescofarina can you please try to re-run the experiments when you have some time?

**Comment by francescofarina** [comment]
I've run the experiments with the new changes but I get very close results for the compiled and non-compiled versions. As you say, there's a chance it's stil not getting compiled IMO. Shall we check whether the loss/accuracy results we get make sense?

**Comment by TristanBilot** [comment]
> Shall we check whether the loss/accuracy results we get make sense?

We can check, but we first have to be sure it's compiled, is there any way to be sure @awni?

**Comment by awni** [comment]
If you do `@mx.compile` it will definitely be compiled. The challenges are two fold:

1. Making sure it is only compiled once. To do that you have to avoid changing constants and inputs sizes and data types
2. Making sure that you are getting the right results (e.g. the data class issue from above).

For 2. I am going to send a fix so it should throw from now on.  For 1. it is a bit tricky, maybe we have a compile debug mode that issues something everytime a function gets compiled. But I would not want that to be the default behavior.

**Comment by thegodone** [comment]
Hi all, I was trying to compile the graph too and have this error

````
 ---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_33902/1786098883.py](http://localhost:8889/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_33902/1786098883.py) in <cell line: 0>()
     72     logger.on_epoch_start(e,optimizer )
     73 
---> 74     loss, train_mse, test_mse, train_rmse, test_rmse = epoch()
     75     logger.on_epoch_end(
     76         e,

[/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_33902/1786098883.py](http://localhost:8889/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_33902/1786098883.py) in epoch()
     58 
     59 def epoch():
---> 60     loss = train(train_loader)
     61     train_mse = test(train_loader)
     62     test_mse = test(test_loader)

[/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_33902/1786098883.py](http://localhost:8889/var/folders/4w/xmf8nmhs51j4vjsttzcxmmm00000gn/T/ipykernel_33902/1786098883.py) in train(train_loader)
     39     loss_sum = 0.0
     40     for graph in train_loader:
---> 41         loss = step(
     42             model=model,
     43             graph=graph,

ValueError: [compile] Function arguments must be trees of arrays or constants (floats, ints, or strings), but received type mlx_graphs.data.batch.GraphDataBatch.
`````
after reading the conversation, compile seems not adapted to graph unpapped version making the compilation not a good choice for graphs, is it still correct as the conversation was in February maybe in between compile as evolve or could evolve ?
What is the memory effect of padding & compile. Did you tests on very large datasets or larger batchsize ?
